### DISCUSSION TITLE
When a PR's title is edited manually, Renovate shouldn't overwrite it when it gets rebased

### DISCUSSION THREAD

**Discussion by nachtien** [renovatebot/renovate#21716] [Suggest an Idea]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

We have to have a Jira ticket associated with every PR (this makes us SOC compliant I think?).  After Renovate creates a PR, we manually prefix a Jira ticket to the title. 

When the PR gets rebased, Renovate resets the PR title back to what it used to be.

I'd like to keep the Jira ticket in the title when the PR gets rebased.

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
The PR title is essential to Renovate to find existing merged or closed PRs. Is the Jira issue number static and therefore could be configured?

**Comment by nachtien** [comment]
Thanks for responding so quickly and reading my issue :) 

I asked if we could have a static Jira issue number and they said "that goes against our SDLC and puts us out of SOC2".

Do you happen to have any other thoughts or suggestions?  

If I understand you correctly, you're saying that modifying the PR title in any way, whether it's rebased or not, could really mess things up?

**Comment by PhilipAbed** [comment]
yes, renovate runs periodically or by webhook triggers of platform(github),
everytime it runs, it needs to pull all the pull requests that are authored by the renovate author/user
if you close the PR or merge it, for example, renovate will be able to know that you want this one ignored/merged, and won't re-open it again and again,

but you changed it's name so renovate won't find it, so it will try to create a new one but it will find the branch name already exists, and the branch name is already part of your PR, 
so what happened is that the Branch was found, and the PR got Re-created, not just overwritten.

this is the expected behavior, next time, change the name + delete/rename the branch, might work then :D

**Comment by nachtien** [comment]
If we change the name + delete/rename, then won't renovate create another PR? 

I'm an Android developer and I really don't know why the design decisions were done, so forgive me for asking :) . Why isn't the PR# referenced instead of the title?

**Comment by rarkins** [comment]
Renovate is designed around using the git repo itself as the storage. Renovate does not keep a list of PR numbers which could get lost of out of sync. So when it needs to ask itself "Has this PR been created before?" then it searches for the branch name + PR title combination. If you mess with that, it won't be able to find the previous PRs.

**Comment by nachtien** [comment]
ohhhhh gotcha, that makes sense!

**Comment by PhilipAbed** [comment]
> If we change the name + delete/rename, then won't renovate create another PR?

yes it will create a new PR 

changing PR title completely is not possible, but u can always play with the title configurations or add extra comments to it

**Comment by PhilipAbed** [comment]
@nachtien you dont need to put the jiraticket in the PR title, 

there's an automatic way to integrate between github and jira, google it (not sure if it works from github to jira worth checking) and
if you want it manually edit it u can add it as a comment

**Comment by nachtien** [comment]
yeah i saw that!  But it would I believe that all our ticket numbers need to be in the title, not a comment :/.  Maybe I'll just program that script to change the title right on merge, or just deal with the duplicate PR's.  When merging in the branch that was renamed, Renovate will see that the version has been updated and close the PR right?

**Comment by nachtien** [comment]
Would it be really difficult to add a configuration that would ignore a prefix? Maybe like some sort of regex string? 

Am I the only one with this problem?

**Comment by rarkins** [comment]
Can you give an example of the before/after PR titles you would want?

**Comment by nachtien** [comment]
I modified Renovate a bit to customize the pr title, but here's a PR thats up:

Update dependency androidx.navigation:navigation-compose from v2.5.1 to v2.5.3

In order for us to merge it in, it would have to look similar to this:

[MXP-1234] Update dependency androidx.navigation:navigation-compose from v2.5.1 to v2.5.3

**Comment by rarkins** [comment]
In this case we'd need to see if the actual PR title _ends with_ the expected title.

**Comment by PhilipAbed** [comment]
> In this case we'd need to see if the actual PR title _ends with_ the expected title.


is there a chance 2 PRS end with the same text?  we should add a flag to support a certain regex template where you can add [MXP-numbers] + pr title... and then we can replace/remove it temporarily when we try to find prs.

we can support this in github only for now, because we might need to change the `findPr` platform function.
maybe we should add a flag for this functionality that works only on github for now?

i thought also we could use the `commitMessagePrefix` but it's not dynamic so not sure

**Reply by youngcm2** [reply]
I ran into the same issue. It would be nice if there could be a separator or pattern to filter out an update to the title.
We run a GitHub action that looks at the PR number to see if there is a corresponding Jira Issue. If not, we append " | [PROJ-9787]" to the PR title.

**Comment by elchenberg** [comment]
I guess this is the place where the bot reverts changes to the pr title: <https://github.com/renovatebot/renovate/blob/2949e135a91b7fbd7599b943582cf721b9c19358/lib/modules/platform/codecommit/index.ts#L420-L422>

Now I am wondering (rather naively) if it might be possible to add `skipPrTitleUpdate: bool` as a new feature to the config (e.g. renovate.json) and then the `updatePr` function would do a check before it updates the title.

```js
  if (title && !skipPrTitleUpdate && cachedPr?.title !== title) {
    await client.updatePrTitle(`${prNo}`, title);
  }
```

**Reply by rarkins** [reply]
Possibly we could do this:
- Store in metadata either the original PR title or simply a hash of it
- If we detect that the PR doesn't match what we expect, check whether it matches the original title
- If the PR title matches the original title then we're free to change it
- Otherwise we assume the PR title was changed by the user, and they want to keep it that way

The main problem for this is if you modify a PR title and then close/ignore the PR. In this case Renovate would not be able to "find" the closed PR and will reopen it, without a lot of additional logic to parse the metadata.

**Reply by elchenberg** [reply]
Understood. Thank you!

**Comment by gzmarstone** [comment]
I would like this feature, we use bitbucket cloud and want some way to link a particular renovate PR to a JIRA ticket and have it not be removed.  The ways that I have seen that jira links PRs to bitbucket is via commit messages or via the PR title/body.  We have tried adding it in the comments and the link does not occur.  Since both the PR Title and Body can be updated by renovate we loose the link if it is updated by renovate - this leads to a weird looking status in jira as it shows that it was once linked, but does not show that the PR was ever deployed (if it was later deployed after renovate removed the JIRA ticket from the body or title).

**Comment by rarkins** [comment]
Converted to a feature request: https://github.com/renovatebot/renovate/issues/26833
